test/UniswapV2PairTest.sol

```diff
@@ -104,38 +140,84 @@ contract UniswapV2PairTest is Test {
     }
 
     function testOptimisticTestCase3() public {
-        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(5), expandTo18Decimals(1));
+        runOptimisticSwapTestCase(997000000000000000, expandTo18Decimals(5), expandTo18Decimals(5));
     }
 
     function testOptimisticTestCase4() public {
-        runOptimisticSwapTestCase(expandTo18Decimals(1), expandTo18Decimals(5), expandTo18Decimals(5), 1003009027081243732);
+        runOptimisticSwapTestCase(expandTo18Decimals(1), expandTo18Decimals(5), expandTo18Decimals(5));
+    }
+
+    function testOptimisticFuzzTestCase(uint outputVal, uint token0Val, uint token1Val) public {
+        (uint outputAmount, uint token0Amount, uint token1Amount) = generateOptimisticFuzzCase(outputVal, token0Val, token1Val);
+        runOptimisticSwapTestCase(outputAmount, token0Amount, token1Amount);
+    }
+
+    function generateOptimisticFuzzCase(uint outputAmountSeed, uint token0AmountSeed, uint token1AmountSeed) private
+            returns (uint _outputAmount, uint _token0Amount, uint _token1Amount) {
+
+        // Token 0 needs at least: A) one output amount to take out and B) one remaining, otherwise all removals will fail:
+        uint minToken0Amount = 2;
+        // Token 0 needs to leave at least 2 in the global supply for an input transfer that will be one or greater with fees (beyond one) taken out:
+        uint maxToken0Amount = MAXIMUM_SUPPLY - 2;
+        _token0Amount = mapSeedToRange(token0AmountSeed, minToken0Amount, maxToken0Amount);
+
+        // The liquidity between the token amounts needs to be 1001 or greater or else mint will fail:
+        uint minToken1Amount = divCeil((MINIMUM_LIQUIDITY + 1) ** 2, _token0Amount);
+        _token1Amount = mapSeedToRange(token1AmountSeed, minToken1Amount, MAXIMUM_SUPPLY);
+
+        // We need to make sure there is enough liquidity to have one remaining, so max liquidity-based output is always one less than token amount:
+        uint maxOutputAmountLiquidity = _token0Amount - 1;
+        // We must also ensure enough tokens remain in global supply to allow the input transfer (when incorporating the 0.03% fee, rounding down):
+        uint maxOutputAmountTotalSupply = ((MAXIMUM_SUPPLY - _token0Amount) * 997) / 1000;
+        _outputAmount = mapSeedToRange(outputAmountSeed, 1, min(maxOutputAmountLiquidity, maxOutputAmountTotalSupply));
     }
 
-    function runOptimisticSwapTestCase(uint outputAmount, uint token0Amount, uint token1Amount, uint inputAmount) private {
+    function runOptimisticSwapTestCase(uint outputAmount, uint token0Amount, uint token1Amount) private {
         addLiquidity(token0Amount, token1Amount);
+        uint inputAmount = computeOptimisticInputAmount(outputAmount);
         token0.transfer(address(pair), inputAmount);
 
-        vm.expectRevert("UniswapV2: K");
+        // The reason we revert may change depending on the relative values of the output amount and available liquidity:
```

Comment:
This was a corner case I just got to the bottom of in my last pass. I'd been thinking output and token0 values of 1 and 2 respectively were causing liquidity issues in the core logic, but it was just the reason we were reverting as expected was different (2 and 2 correctly declared INSUFFICIENT_LIQUIDITY, but all other cases off the border declared invalid K). This let me reduce the minimum token0 amount once I realized it...